core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala

```diff
@@ -0,0 +1,384 @@
+
+  startWith(
+    Offline,
+    InvokerInfo(
+      new RingBuffer[Boolean](InvokerHealthManager.bufferSize),
+      memory = MemoryInfo(instanceId.userMemory.toMB, 0, 0)))
+
+  when(Offline) {
+    case Event(GracefulShutdown, _: InvokerInfo) =>
+      logging.warn(this, "Received a graceful shutdown flag, stopping the invoker.")
+      stay
+
+    case Event(Enable, _) =>
+      InvokerHealthManager.prepare(entityStore, instanceId).map { _ =>
+        startTestAction(self)
+      }
+      goto(Unhealthy)
+  }
+
+  when(Unhealthy) {
+    case Event(ContainerRemoved, _) =>
+      healthActionProxy = None
+      startTestAction(self)
+      stay
+
+    case Event(msg: FailureMessage, _) =>
+      logging.error(this, s"invoker${instanceId}, status:${stateName} got a failure message: ${msg}")
+      stay
+
+    case Event(ContainerCreationFailed(_), _) =>
+      stay
+  }
+
+  when(Healthy) {
+    case Event(msg: FailureMessage, _) =>
+      logging.error(this, s"invoker${instanceId}, status:${stateName} got a failure message: ${msg}")
+      goto(Unhealthy)
+  }
+
+  whenUnhandled {
+    case Event(_: Initialized, _) =>
+      // Initialized messages sent by ContainerProxy for HealthManger
+      stay()
+
+    case Event(ContainerRemoved, _) =>
+      // Drop messages sent by ContainerProxy for HealthManger
+      healthActionProxy = None
+      stay()
+
+    case Event(GracefulShutdown, _) =>
+      self ! GracefulShutdown
+      goto(Offline)
+
+    case Event(healthMsg: HealthMessage, data: InvokerInfo) =>
+      if (stateName != Offline) {
+        handleHealthMessage(healthMsg.state, data.buffer)
+      } else {
+        stay
+      }
+
+    case Event(memoryInfo: MemoryInfo, data: InvokerInfo) =>
+      publishHealthStatusAndStay(stateName, data.copy(memory = memoryInfo))
+
+    // in case of StatusRuntimeException: NOT_FOUND: etcdserver: requested lease not found, we need to get the lease again.
+    case Event(t: FailureMessage, _) =>
+      logging.error(this, s"Failure happens, restart InvokerHealthManager: ${t}")
+      goto(Offline)
+  }
+
+  // It is important to note that stateName and the stateData in onTransition callback refer to the previous one.
+  // We should access to the next data with nextStateData
+  onTransition {
+    case Offline -> Unhealthy =>
+      publishHealthStatusAndStay(Unhealthy, nextStateData)
+
+    case Healthy -> Unhealthy =>
+      unstashAll()
+      transid.mark(
+        this,
+        LoggingMarkers.LOADBALANCER_INVOKER_STATUS_CHANGE(Unhealthy.asString),
+        s"invoker${instanceId.toInt} is unhealthy",
+        akka.event.Logging.WarningLevel)
+      startTestAction(self)
+      publishHealthStatusAndStay(Unhealthy, nextStateData)
+
+    case _ -> Healthy =>
+      logging.info(this, s"invoker became healthy, stop health action proxy.")
+      unstashAll()
+      stopTestAction()
+
+      publishHealthStatusAndStay(Healthy, nextStateData)
+
+    case Offline -> Offline =>
```

Comment:
nit: can't this be removed? you can just leave out state transitions that don't do anything for this partial function

Reply:
+1

Reply:
Already fixed